### DISCUSSION TITLE
How to implement whitespace and line feed processing in the output data？

### DISCUSSION THREAD

**Discussion by Apricity-yu** [ajaxorg/ace#5172] [Q&A]
### Describe the feature

How can I identify the space and carriage return problems when displaying data
When the data is returned by the interface, what format should be processed to assign it to the setValue method, display whitespace recognition and line wrapping in ace-build

### Use Case

For example:  function transformPayload (rawData) {var s={};s.test= rawData;return s;}

The current data is shown in a row in ace-builds, which bothers me, and I expect the format to be as follows.
The format I expect is as follows：
function transformPayload (rawData) {
var s={};
s.test= rawData;
return s;
}

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### ACE version used

^1.20.0

**Comment by akoreman** [comment]
Thanks for reaching out, unfortunately, it's not completely clear to me what problem you're facing. Could you expand a bit more on what you are trying to do and where you're facing problems?

**Comment by InspiredGuy** [comment]
Hi @Apricity-yu, I am not sure I understood your request correctly, but if you want to know how to make Ace automatically format the text you provide to it via `EditSession.setValue` - unfortunately, this is not possible at the moment. If your question is related to how to put newlines in the text so Ace renders newlines correctly - you could get the newline character used in the document using [Document.getNewlineCharacter](https://github.com/ajaxorg/ace/blob/1bc5d9929ca4dd867480a74381973f632154bc6e/src/document.js#LL74C5-L74C24) and insert it in the relevant places of the string you pass to `setValue`.

**Comment by InspiredGuy** [comment]
You could also try to use existing [beautify](https://github.com/ajaxorg/ace/blob/master/src/ext/beautify.js) extension to enable the command to format the code.
